sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala

```diff
@@ -132,11 +130,12 @@ case class BatchScanExec(
 
       outputPartitioning match {
         case p: KeyGroupedPartitioning =>
-          val partitionMapping = finalPartitions.map(s =>
-            s.head.asInstanceOf[HasPartitionKey].partitionKey() -> s).toMap
+          val partitionMapping = finalPartitions
+            .map(s => InternalRowComparableWrapper(s.head, p.expressions) -> s).toMap
           finalPartitions = p.partitionValues.map { partValue =>
             // Use empty partition for those partition values that are not present
-            partitionMapping.getOrElse(partValue, Seq.empty)
```

Comment:
`Map` operation needs to rely on `equals` method, change to use `InternalRowComparableWrapper`